README.md
# mongoose-autopopulate

Always `populate()` certain fields in your mongoose schemas

[![Build Status](https://travis-ci.org/mongodb-js/mongoose-autopopulate.svg?branch=master)](https://travis-ci.org/mongodb-js/mongoose-autopopulate)
[![Coverage Status](https://coveralls.io/repos/mongodb-js/mongoose-autopopulate/badge.svg?branch=master)](https://coveralls.io/r/mongodb-js/mongoose-autopopulate?branch=master)

[Read the docs here](http://plugins.mongoosejs.io/plugins/autopopulate).

**Note:** population is a powerful feature, but it has limitations and
helps you get away with poor schema design.  In particular, it is usually
bad MongoDB schema design to include arrays that grow without bound in
your documents. Do not include a constantly-growing array of ObjectIds
in your schema - your data will become unwieldy as the array grows and
you will eventually hit the [16 MB document size limit](http://docs.mongodb.org/manual/reference/limits/#BSON-Document-Size).
In general, think carefully when designing your schemas.

# Usage

The `mongoose-autopopulate` module exposes a single function that you can
pass to [Mongoose schema's `plugin()` function](https://mongoosejs.com/docs/api.html#schema_Schema-plugin).

```javascript
const schema = new mongoose.Schema({
  populatedField: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'ForeignModel',
    // The below option tells this plugin to always call `populate()` on
    // `populatedField`
    autopopulate: true
  }
});
schema.plugin(require('mongoose-autopopulate'));
```

Only apply this plugin to top-level schemas. Don't apply this plugin to child schemas.

```javascript
// Don't do `nestedSchema.plugin(require('mongoose-autopopulate'))`.
// You only need to add mongoose-autopopulate to top-level schemas.
const nestedSchema = mongoose.Schema({
  child: { type: Number, ref: 'Child', autopopulate: true }
});
const topSchema = mongoose.Schema({ nested: nestedSchema });
topSchema.plugin(require('mongoose-autopopulate'));
```


History.md
1.1.0 / 2023-10-24
==================
 * feat: support mongoose 8

1.0.1 / 2023-05-05
==================
 * fix: support nested discriminators #113 #107

1.0.0 / 2023-02-27
==================
 * feat: support mongoose 7

1.0.0-rc0 / 2023-02-24
======================
 * feat: support mongoose 7

0.17.1 / 2022-01-03
===================
 * fix: handle infinite schema recursion #106 #101 [IslandRhythms](https://github.com/IslandRhythms)

0.17.0 / 2022-12-28
===================
 * feat: add support for nested autopopulated virtuals #105 #104 [Danchoys](https://github.com/Danchoys)

0.16.1 / 2022-04-04
===================
 * fix: support subdocuments in Mongoose 6.1.x #94

0.16.0 / 2021-09-22
===================
 * feat: add native type definitions #92 #91 [TheVaan](https://github.com/TheVaan)

0.15.0 / 2021-09-06
===================
 * BREAKING CHANGE: use Mongoose 6, drop support for Mongoose 5 #89 #88

0.14.0 / 2021-05-14
===================
 * feat: add `functions` option to list out functions to register hooks on

0.13.0 / 2021-04-15
===================
 * fix: autopopulate nested array fields containing embedded discriminator #82
 * feat: autopopulate embedded discriminators #82

0.12.2 / 2020-04-19
===================
 * fix: avoid error in post('save') handler when this plugin is registered on a child schema #10
 * docs: explain that this plugin should only be registered on top-level schemas

0.12.1 / 2020-03-29
===================
 * fix: handle autopopulate within nested document array when top-level array is empty #70

0.12.0 / 2020-02-05
===================
 * feat: autopopulate discriminators post `find()`, `findOne()`, and `findOneAndUpdate()` #26

0.11.0 / 2020-01-24
===================
 * feat: allow override of maxDepth param via query options #62 [jmikrut](https://github.com/jmikrut)

0.10.0 / 2019-12-20
===================
 * feat: autopopulate paths after `save()` if they aren't already populated #8

0.9.1 / 2019-01-02
==================
 * docs: add note about selectPopulatedPaths option #50

0.9.0 / 2018-11-08
==================
 * feat: support turning on autopopulate with lean #48 #27 #14

0.8.2 / 2018-10-10
==================
 * docs: link to new plugins site on plugins.mongoosejs.io

0.8.1 / 2018-09-02
==================
 * fix: call function with options and include refPath in options #15

0.8.0 / 2018-07-01
==================
 * fix: add support for findOneAndUpdate() with autopopulate #42

0.7.0 / 2018-05-10
==================
 * BREAKING CHANGE: drop support for Node.js < 4.0.0
 * feat: add `maxDepth` option, set to 10 by default #37 #20 #11

0.5.0 / 2016-11-29
==================
 * feat: support lean #27 #14 [siboulet](https://github.com/siboulet)
 * feat: support virtual autopopulate #24 [siboulet](https://github.com/siboulet)

0.4.0 / 2015-09-03
==================
 * added; support for arrays #9 [rerthal](https://github.com/rerthal)

0.3.0 / 2015-07-30
==================
 * added; support of autopopulate in nested schemas #5 [eloytoro](https://github.com/eloytoro)
 * fixed; changed mongoose peer dependency to 4.x #4

0.2.0 / 2015-06-27
==================
 * added; support for nested schemas #3 [siboulet](https://github.com/siboulet)


test/unit.js
'use strict';

const plugin = require('../');
const assert = require('assert');

describe('mongoose-autopopulate:unit', function() {
  let schemaStub;
  let paths = [];

  it('populates when paths autopopulate option is true', function() {
    paths = [{ name: 'test', options: { options: { autopopulate: true } } }];
    schemaStub = createSchemaStub(paths);
    plugin(schemaStub);
    assert.equal(schemaStub.pre.calls.length, 5);

    const pathsToPopulate = schemaStub.pre.calls[0].handler.call({});
    assert.equal(pathsToPopulate.length, 1);
    assert.equal(pathsToPopulate[0].path, 'test');
  });

  it('ignores when paths autopopulate option is falsy', function() {
    paths = [
      { name: 'test', options: { options: { autopopulate: true } } },
      { name: 'test2', options: { options: { autopopulate: false } } },
      { name: 'test3', options: { options: { autopopulate: null } } }
    ];
    schemaStub = createSchemaStub(paths);
    plugin(schemaStub);
    assert.equal(schemaStub.pre.calls.length, 5);

    const pathsToPopulate = schemaStub.pre.calls[0].handler.call({});
    assert.equal(pathsToPopulate.length, 1);
    assert.equal(pathsToPopulate[0].path, 'test');
  });

  it('merges options when autopopulate option is object', function() {
    paths = [
      {
        name: 'test',
        options: {
          options: {
            autopopulate: { select: 'name' }
          }
        }
      }
    ];
    schemaStub = createSchemaStub(paths);

    plugin(schemaStub);
    assert.equal(schemaStub.pre.calls.length, 5);

    const pathsToPopulate = schemaStub.pre.calls[0].handler.call({});
    assert.equal(pathsToPopulate.length, 1);
    assert.deepEqual(pathsToPopulate[0],
      { path: 'test', select: 'name', options: { _depth: 1, maxDepth: 10 } });
  });

  it('can execute function options', function() {
    paths = [
      {
        name: 'test',
        options: {
          options: {
            autopopulate: function() {
              return true;
            }
          }
        }
      }
    ];
    schemaStub = createSchemaStub(paths);
    plugin(schemaStub);
    assert.equal(schemaStub.pre.calls.length, 5);

    const pathsToPopulate = schemaStub.pre.calls[0].handler.call({});
    assert.equal(pathsToPopulate.length, 1);
    assert.deepEqual(pathsToPopulate[0].path, 'test');
  });

  it('augments populate options when autopopulate returns object', function() {
    paths = [
      {
        name: 'test',
        options: {
          options: {
            autopopulate: function() {
              return { select: 'name' };
            }
          }
        }
      }
    ];
    schemaStub = createSchemaStub(paths);

    plugin(schemaStub);
    assert.equal(schemaStub.pre.calls.length, 5);

    const pathsToPopulate = schemaStub.pre.calls[0].handler.call({});
    assert.equal(pathsToPopulate.length, 1);
    assert.deepEqual(pathsToPopulate[0],
      { path: 'test', select: 'name', options: { maxDepth: 10, _depth: 1 } });
  });

  it('doesnt populate when autopopulate function returns falsy', function() {
    paths = [
      {
        name: 'test',
        options: {
          options: {
            autopopulate: function() {
              return false;
            }
          }
        }
      }
    ];
    schemaStub = createSchemaStub(paths);
    plugin(schemaStub);
    assert.equal(schemaStub.pre.calls.length, 5);

    const pathsToPopulate = schemaStub.pre.calls[0].handler.call({});
    assert.equal(pathsToPopulate.length, 0);
  });

  it('handles nested schemas', function() {
    const nestedPath = {
      name: 'test',
      options: {
        options: {
          autopopulate: true
        }
      }
    };

    const topLevel = {
      name: 'nested',
      options: {
        schema: createSchemaStub([nestedPath])
      }
    };

    const schema = createSchemaStub([topLevel]);

    plugin(schema);
    assert.equal(schema.pre.calls.length, 5);

    const pathsToPopulate = schema.pre.calls[0].handler.call({});
    assert.deepEqual(pathsToPopulate[0].path,
      'nested.test');
  });

  it('handles top-level populated virtuals', () => {
    schemaStub = createSchemaStub([]);
    schemaStub.virtuals = {
      test: {
        options: { autopopulate: true }
      }
    };

    plugin(schemaStub);
    assert.equal(schemaStub.pre.calls.length, 5);

    const pathsToPopulate = schemaStub.pre.calls[0].handler.call({});
    assert.equal(pathsToPopulate.length, 1);
    assert.equal(pathsToPopulate[0].path, 'test');
  });

  it('handles nested populated virtuals', () => {
    const nestedSchema = createSchemaStub([]);
    nestedSchema.virtuals = {
      test: {
        options: { autopopulate: true }
      }
    };

    const topLevel = {
      name: 'nested',
      options: {
        schema: nestedSchema
      }
    };

    const schema = createSchemaStub([topLevel]);

    plugin(schema);
    assert.equal(schema.pre.calls.length, 5);

    const pathsToPopulate = schema.pre.calls[0].handler.call({});
    assert.equal(pathsToPopulate.length, 1);
    assert.equal(pathsToPopulate[0].path, 'nested.test');
  });
});

function createSchemaStub(paths) {
  const schemaStub = {};
  schemaStub.pre = function(func, handler) {
    schemaStub.pre.calls.push({ func: func, handler: handler });
    return schemaStub;
  };
  schemaStub.pre.calls = [];
  schemaStub.post = function(func, handler) {
    schemaStub.post.calls.push({ func: func, handler: handler });
    return schemaStub;
  };
  schemaStub.post.calls = [];
  schemaStub.eachPath = function(handler) {
    paths.forEach(function(path) {
      handler(path.name, path.options);
    });
  };

  return schemaStub;
}


test/integration.js
'use strict';

const assert = require('assert');
const autopopulate = require('../');
const mongoose = require('mongoose');

const Schema = mongoose.Schema;
const ObjectId = mongoose.Schema.Types.ObjectId;

describe('mongoose-autopopulate plugin', function() {
  let Band;
  let Person;

  before(async function() {
    await mongoose.connect('mongodb://127.0.0.1:27017/autopopulate');

    const personSchema = new Schema({ name: String, birthName: String });
    const bandSchema = new Schema({
      name: String,
      lead: { type: ObjectId, ref: 'people', autopopulate: true },
      members: [{ type: ObjectId, ref: 'people', autopopulate: true }]
    });
    bandSchema.plugin(autopopulate);
    Person = mongoose.model('people', personSchema, 'people');
    Band = mongoose.model('band', bandSchema, 'bands');
  });

  beforeEach(async function() {
    const axl = {
      name: 'Axl Rose',
      birthName: 'William Bruce Rose, Jr.'
    };

    const gnr = { name: 'Guns N\' Roses' };

    await Person.deleteMany({});
    await Band.deleteMany({});
    const doc = await Person.create(axl);
    assert.ok(doc);
    gnr.lead = doc._id;
    gnr.members = [doc._id];
    await Band.create(gnr);
  });

  after(function() {
    return mongoose.disconnect();
  });

  /**
   * Suppose you have two collections, "people" and "bands". The `People` model
   * looks like this:
   *
   * ```javascript
   * var personSchema = new Schema({ name: String, birthName: String });
   * Person = mongoose.model('people', personSchema, 'people');
   * ```
   *
   * Suppose your "people" collection has one document:
   *
   * ```javascript
   * {
   *   name: 'Axl Rose',
   *   birthName: 'William Bruce Rose, Jr.',
   *   _id: '54ef3f374849dcaa649a3abc'
   * };
   * ```
   *
   * And your "bands" collection has one document:
   *
   * ```javascript
   * {
   *   _id: '54ef3f374849dcaa649a3abd',
   *   name: "Guns N' Roses",
   *   lead: '54ef3f374849dcaa649a3abc',
   *   members: ['54ef3f374849dcaa649a3abc']
   * }
   * ```
   *
   *  You can set the `autopopulate` option for the `lead` field.
   *  This means that, every time you call `find()` or `findOne()`,
   *  `mongoose-autopopulate` will automatically call `.populate('lead')`
   *  for you.
   */
  it('supports an autopopulate option in schemas', async function() {
    const bandSchema = new Schema({
      name: String,
      lead: { type: ObjectId, ref: 'people', autopopulate: true }
    });
    bandSchema.plugin(autopopulate);

    const Band = mongoose.model('band3', bandSchema, 'bands');
    const doc = await Band.findOne({ name: 'Guns N\' Roses' });
    assert.equal('Axl Rose', doc.lead.name);
    assert.equal('William Bruce Rose, Jr.', doc.lead.birthName);
  });


  /**
   *  `mongoose-autopopulate` also works on arrays.
   */
  it('supports document arrays', async function() {
    const bandSchema = new Schema({
      name: String,
      members: [{ type: ObjectId, ref: 'people', autopopulate: true }]
    });
    bandSchema.plugin(autopopulate);

    const Band = mongoose.model('band4', bandSchema, 'bands');
    const doc = await Band.findOne({ name: 'Guns N\' Roses' });
    assert.equal('Axl Rose', doc.members[0].name);
    assert.equal('William Bruce Rose, Jr.', doc.members[0].birthName);
  });

  /**
   *  By default, Mongoose 5.x automatically projects in populated properties.
   *  That means you need a little extra work to exclude autopopulated fields.
   *  Either explicitly [deselect the path](https://mongoosejs.com/docs/api.html#query_Query-select)
   *  in your projection, or set the [`selectPopulatedPaths` schema option](https://mongoosejs.com/docs/guide.html#selectPopulatedPaths)
   *  to `false`.
   */
  it('has a couple caveats with projections', async function() {
    // Mongoose adds `members: 1` and `lead: 1` to the projection
    let band = await Band.findOne().select({ name: 1 });
    assert.equal(band.members[0].name, 'Axl Rose');
    assert.equal(band.lead.name, 'Axl Rose');

    // You can also tell Mongoose to not project in populated paths by default
    // using the `selectPopulatedPaths` schema option.
    const newSchema = Band.schema.clone();

    newSchema.options.selectPopulatedPaths = false;
    const Band2 = mongoose.model('Band2', newSchema, 'bands');

    band = await Band2.findOne().select({ name: 1 });
    assert.ok(!band.members);
    assert.ok(!band.lead);
  });

  /**
   *  Advanced users of `populate()` may want to specify additional
   *  options, such as selecting fields. If you set the `autopopulate`
   *  option to an object, `mongoose-autopopulate` will merge the object
   *  into populate options. The `findOne()` below is equivalent to
   *  `Band.findOne({ name: "Guns N' Roses" }).populate({ path: 'lead', select: 'name });`
   */
  it('can specify an options argument', async function() {
    const bandSchema = new Schema({
      name: String,
      lead: { type: ObjectId, ref: 'people', autopopulate: { select: 'name' } }
    });
    bandSchema.plugin(autopopulate);

    const Band = mongoose.model('band5', bandSchema, 'bands');
    const doc = await Band.findOne({ name: 'Guns N\' Roses' });
    assert.equal('Axl Rose', doc.lead.name);
    assert.ok(!doc.lead.birthName);
  });

  /**
   *  You can also set the `autopopulate` option to be a function.
   *  Then `mongoose-autopopulate` will call the function with
   *  the query object as the context and use the return value.
   *  The below `populate()` uses the same options as the previous
   *  example.
   */
  it('can specify a function that returns options', async() => {
    let numCalls = 0;
    const optionsFunction = function() {
      ++numCalls;
      return { select: 'name' };
    };

    const bandSchema = new Schema({
      name: String,
      lead: { type: ObjectId, ref: 'people', autopopulate: optionsFunction }
    });
    bandSchema.plugin(autopopulate);

    const Band = mongoose.model('band6', bandSchema, 'bands');
    const docs = await Band.find({ name: 'Guns N\' Roses' });
    assert.equal(1, docs.length);
    assert.equal(1, numCalls);
    const doc = docs[0];
    assert.equal('Axl Rose', doc.lead.name);
    assert.ok(!doc.lead.birthName);
  });

  /**
   *  If you set the `autopopulate` option to `false` on a query, autopopulate
   *  will be disabled. This is handy if you want to autopopulate by default,
   *  but opt-out for special cases.
   */
  it('can disable autopopulate for individual queries', async() => {
    const bandSchema = new Schema({
      name: String,
      lead: { type: ObjectId, ref: 'people', autopopulate: true }
    });
    bandSchema.plugin(autopopulate);

    const Band = mongoose.model('band7', bandSchema, 'bands');
    const doc = await Band.findOne({ name: 'Guns N\' Roses' }, {}, { autopopulate: false });
    assert.ok(doc.lead instanceof mongoose.Types.ObjectId);
    assert.ok(!doc.populated('lead'));
  });


  /**
   *  Say you have a model `User` that has the autopopulate plugin and you're
   *  populating users from a different model. To disable autopopulate, you
   *  need to set `autopopulate: false` as a populate option, not a query
   *  option.
   */

  it('can disable autopopulate in `populate()` options', async function() {
    const userSchema = new Schema({
      name: String,
      friends: [{
        type: Schema.Types.ObjectId,
        ref: 'User',
        autopopulate: { maxDepth: 2 }
      }]
    });
    userSchema.plugin(autopopulate);

    const responseSchema = new Schema({
      user: {
        type: Schema.Types.ObjectId,
        ref: 'User'
      }
    });

    const User = mongoose.model('User', userSchema);
    const Response = mongoose.model('Response', responseSchema);

    const axl = new User({ name: 'Axl' });
    const slash = new User({ name: 'Slash', friends: [axl._id] });
    axl.friends.push(slash._id);

    await Promise.all([axl.save(), slash.save()]);
    let r = await Response.create({ user: axl._id });

    r = await Response.findById(r._id).
      // Because `User` is the foreign model, you need to disable autopopulate
      // in the populate options below, not the query options
      populate({ path: 'user', options: { autopopulate: false } });
    // acquit:ignore:start
    assert.equal(r.user.name, 'Axl');
    assert.ok(r.user.friends[0] instanceof mongoose.Types.ObjectId);
    // acquit:ignore:end
  });

  /**
   *  Setting the [Mongoose `lean` option](https://mongoosejs.com/docs/api.html#query_Query-lean)
   *  will disable autopopulate for all paths, _unless_ you add `autopulate: true`
   *  to your `lean` option.
   */

  it('requires an option to work with lean', async function() {
    let band = await Band.findOne().lean();
    // Won't autopopulate because `lean()` is set
    assert.ok(band.lead instanceof mongoose.Types.ObjectId);

    // To turn on `autopopulate` with lean, use `lean({ autopulate: true })`
    band = await Band.findOne().lean({ autopopulate: true });
    assert.equal(band.lead.name, 'Axl Rose');
  });

  /**
   *  Recursive populate can lead to messy infinite recursion, so this plugin
   *  supports a `maxDepth` option that limits how deep recursive population
   *  will go. The `maxDepth` option is 10 by default
   */
  it('can limit the depth using `maxDepth`', async function() {
    const accountSchema = new mongoose.Schema({
      name: String,
      friends: [{
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Account',
        // This is a recursive relationship, `friends` points to a list
        // of accounts. If we didn't limit the depth, this would result
        // in infinite recursion!
        autopopulate: { maxDepth: 2 }
      }]
    });
    accountSchema.plugin(autopopulate);

    const Account = mongoose.model('Account', accountSchema);

    const axl = new Account({ name: 'Axl' });
    const slash = new Account({ name: 'Slash', friends: [axl._id] });
    axl.friends.push(slash._id);

    await axl.save();
    await slash.save();

    const doc = await Account.findById(axl._id);

    assert.equal(doc.friends[0].name, 'Slash');
    assert.equal(doc.friends[0].friends[0].name, 'Axl');
    // Only populate 2 levels deep, 3rd level will still be an `_id`
    assert.equal(doc.friends[0].friends[0].friends[0].toString(),
      slash._id.toHexString());
  });

  /**
   *  By default, autopopulate applies to the results of `find()`, `findOne()`,
   *  `findOneAndUpdate()`, and `save()`. You can pick which functions
   *  you want autopopulate to handle using the `functions` option. For example,
   *  the below code disables autopopulating on `save()`.
   */
  it('can pass a list or regular expression of functions to apply hooks to', async function() {
    const bandSchema = new Schema({
      name: String,
      lead: { type: ObjectId, ref: 'people', autopopulate: true }
    });
    bandSchema.plugin(autopopulate, {
      // Apply this plugin to all functions except for `save()`
      functions: ['find', 'findOne', 'findOneAndUpdate']
    });

    const Band = mongoose.model('band8', bandSchema, 'bands');

    let band = await Band.findOne({ name: 'Guns N\' Roses' });
    assert.ok(band.populated('lead'));

    band = await Band.findOne({ name: 'Guns N\' Roses' }).setOptions({ autopopulate: false });
    assert.ok(!band.populated('lead'));
    // `save()` doesn't autopopulate
    await band.save();
    assert.ok(!band.populated('lead'));
  });
  it('supports findOneAndDelete', async function() {
    const bandSchema = new Schema({
      name: String,
      lead: { type: ObjectId, ref: 'people', autopopulate: true }
    });
    bandSchema.plugin(autopopulate, {
      functions: ['findOneAndDelete']
    });

    const Band = mongoose.model('band9', bandSchema, 'bands');

    const band = await Band.findOneAndDelete({ name: 'Guns N\' Roses' });
    assert.ok(band.populated('lead'));
  });
  it('supports findOneAndReplace', async function() {
    const bandSchema = new Schema({
      name: String,
      lead: { type: ObjectId, ref: 'people', autopopulate: true }
    });
    bandSchema.plugin(autopopulate, {
      functions: ['findOneAndReplace']
    });

    const Band = mongoose.model('band10', bandSchema, 'bands');
    const p = await Person.create({ name: 'Test', birthName: 'Testerson' });
    const band = await Band.findOneAndReplace({ name: 'Guns N\' Roses' }, { name: 'Testing123', lead: p._id });
    const newBand = await Band.findOne({ name: 'Testing123' });
    assert.ok(band.populated('lead'));
    assert.equal(newBand.name, 'Testing123');
  });
});


index.js
'use strict';

module.exports = function autopopulatePlugin(schema, options) {
  const pathsToPopulate = getPathsToPopulate(schema);

  let testFunction = () => true;
  if (options != null && (Array.isArray(options.functions) || options.functions instanceof RegExp)) {
    let _functions = options.functions;
    if (Array.isArray(_functions)) {
      _functions = new Set(_functions);

      testFunction = v => _functions.has(v);
    } else {
      testFunction = v => _functions.test(v);
    }
  }

  const autopopulateHandler = function(filter) {
    const finalPaths = [];

    if (this._mongooseOptions &&
        this._mongooseOptions.lean &&
        // If lean and user didn't explicitly do `lean({ autopopulate: true })`,
        // skip it. See gh-27, gh-14, gh-48
        !this._mongooseOptions.lean.autopopulate) {
      return;
    }

    const options = this.options || {};
    if (options.autopopulate === false) {
      return;
    }

    if (options.autopopulate && options.autopopulate.maxDepth) {
      options.maxDepth = options.autopopulate.maxDepth;
    }

    const depth = options._depth != null ? options._depth : 0;
    if (options.maxDepth > 0 && depth >= options.maxDepth) {
      return;
    }

    const numPaths = pathsToPopulate.length;
    for (let i = 0; i < numPaths; ++i) {
      pathsToPopulate[i].options = pathsToPopulate[i].options || {};
      if (typeof filter === 'function' && !filter(pathsToPopulate[i].options)) {
        continue;
      }
      pathsToPopulate[i].options.options = pathsToPopulate[i].options.options || {};
      const newOptions = { _depth: depth + 1 };
      if (options.maxDepth) newOptions.maxDepth = options.maxDepth;
      Object.assign(pathsToPopulate[i].options.options, newOptions);
      const optionsToUse = processOption.call(this,
        pathsToPopulate[i].autopopulate, pathsToPopulate[i].options);
      if (optionsToUse) {
        // If `this` is a query, population chaining is allowed.
        // If not, add it to an array for single population at the end.
        if (this.constructor.name === 'Query') {
          this.populate(optionsToUse);
        } else {
          finalPaths.push(optionsToUse);
        }
      }
    }
    return finalPaths;
  };

  if (testFunction('find')) {
    schema.pre('find', function() { return autopopulateHandler.call(this); });
    schema.post('find', function(res) { return autopopulateDiscriminators.call(this, res); });
  }
  if (testFunction('findOne')) {
    schema.pre('findOne', function() { return autopopulateHandler.call(this); });
    schema.post('findOne', function(res) { return autopopulateDiscriminators.call(this, res); });
  }
  if (testFunction('findOneAndUpdate')) {
    schema.pre('findOneAndUpdate', function() { return autopopulateHandler.call(this); });
    schema.post('findOneAndUpdate', function(res) { return autopopulateDiscriminators.call(this, res); });
  }
  if (testFunction('findOneAndDelete')) {
    schema.pre('findOneAndDelete', function() { return autopopulateHandler.call(this); });
    schema.post('findOneAndDelete', function(res) { return autopopulateDiscriminators.call(this, res); });
  }
  if (testFunction('findOneAndReplace')) {
    schema.pre('findOneAndReplace', function() { return autopopulateHandler.call(this); });
    schema.post('findOneAndReplace', function(res) { return autopopulateDiscriminators.call(this, res); });
  }
  if (testFunction('save')) {
    schema.post('save', function() {
      if (pathsToPopulate.length === 0) {
        return Promise.resolve();
      }
      // Skip for subdocs, because we assume this function only runs on
      // top-level documents.
      if (typeof this.ownerDocument === 'function' && this.$isSubdocument) {
        return Promise.resolve();
      }
      const finalPaths = autopopulateHandler.call(this, options => {
        const pop = this.populated(options.path);
        if (Array.isArray(pop)) {
          const docVal = this.get(options.path);
          return docVal == null ||
            pop.length !== docVal.length ||
            pop.some(v => v == null);
        }
        return true;
      });
      return this.populate(finalPaths);
    });
  }
};

function autopopulateDiscriminators(res) {
  if (res == null) {
    return;
  }
  if (this._mongooseOptions != null && this._mongooseOptions.lean) {
    // If lean, we don't have a good way to figure out the discriminator
    // schema, and so skip autopopulating.
    return;
  }
  if (!Array.isArray(res)) {
    res = [res];
  }

  const discriminators = new Map();
  for (const doc of res) {
    if (doc.constructor.baseModelName != null) {
      const discriminatorModel = doc.constructor;
      const modelName = discriminatorModel.modelName;

      if (!discriminators.has(modelName)) {
        const pathsToPopulate = getPathsToPopulate(discriminatorModel.schema).
          filter(p => !doc.populated(p.options.path));

        discriminators.set(modelName, {
          model: discriminatorModel,
          docs: [],
          pathsToPopulate: pathsToPopulate
        });
      }
      const modelMap = discriminators.get(modelName);
      modelMap.docs.push(doc);
    }
  }

  return Promise.all(Array.from(discriminators.values()).map(modelMap => {
    const pathsToPopulate = modelMap.pathsToPopulate.
      map(p => processOption.call(this, p.autopopulate, p.options)).
      filter(v => !!v);
    return modelMap.model.populate(modelMap.docs, pathsToPopulate);
  }));
}

function getPathsToPopulate(schema) {
  const pathsToPopulate = [];
  const schemaStack = new WeakMap();
  eachPathRecursive(schema, function(pathname, schemaType) {
    let option;
    if (schemaType.options && schemaType.options.autopopulate) {
      option = schemaType.options.autopopulate;
      pathsToPopulate.push({
        options: defaultOptions(pathname, schemaType.options),
        autopopulate: option
      });
    } else if (schemaType.options &&
        schemaType.options.type &&
        schemaType.options.type[0] &&
        schemaType.options.type[0].autopopulate) {
      option = schemaType.options.type[0].autopopulate;
      pathsToPopulate.push({
        options: defaultOptions(pathname, schemaType.options.type[0]),
        autopopulate: option
      });
    }
  }, null, schemaStack);
  return pathsToPopulate;
}

function defaultOptions(pathname, v) {
  const ret = { path: pathname, options: { maxDepth: 10 } };
  if (v.ref != null) {
    ret.model = v.ref;
    ret.ref = v.ref;
  }
  if (v.refPath != null) {
    ret.refPath = v.refPath;
  }
  return ret;
}

function processOption(value, options) {
  switch (typeof value) {
    case 'function':
      return handleFunction.call(this, value, options);
    case 'object':
      return handleObject.call(this, value, options);
    default:
      return handlePrimitive.call(this, value, options);
  }
}

function handlePrimitive(value, options) {
  if (value) {
    return options;
  }
}

function handleObject(value, optionsToUse) {
  // Special case: support top-level `maxDepth`
  if (value.maxDepth != null) {
    optionsToUse.options = optionsToUse.options || {};
    optionsToUse.options.maxDepth = value.maxDepth;
    delete value.maxDepth;
  }
  optionsToUse = Object.assign({}, optionsToUse, value);

  return optionsToUse;
}

function handleFunction(fn, options) {
  const val = fn.call(this, options);
  return processOption.call(this, val, options);
}

function eachPathRecursive(schema, handler, path, schemaStack) {

  if (schemaStack.has(schema)) {
    return;
  }
  if (!path) {
    path = [];
  }
  schemaStack.set(schema, true);

  schema.eachPath(function(pathname, schemaType) {
    path.push(pathname);
    if (schemaType.schema) {
      eachPathRecursive(schemaType.schema, handler, path, schemaStack);

      if (schemaType.schema.discriminators != null) {
        for (const discriminatorName of Object.keys(schemaType.schema.discriminators)) {
          eachPathRecursive(
            schemaType.schema.discriminators[discriminatorName],
            handler,
            path,
            schemaStack
          );
        }
      }
    } else if (schemaType.$isMongooseArray && schemaType.$embeddedSchemaType.$isMongooseArray) {
      while (schemaType != null && schemaType.$isMongooseArray && !schemaType.$isMongooseDocumentArray) {
        schemaType = schemaType.$embeddedSchemaType;
      }
      if (schemaType != null && schemaType.$isMongooseDocumentArray) {
        eachPathRecursive(schemaType.schema, handler, path, schemaStack);

        if (schemaType.schema.discriminators != null) {
          for (const discriminatorName of Object.keys(schemaType.schema.discriminators)) {
            eachPathRecursive(
              schemaType.schema.discriminators[discriminatorName],
              handler,
              path,
              schemaStack
            );
          }
        }
      }
    } else {
      handler(path.join('.'), schemaType);
    }
    path.pop();
  });
  schemaStack.delete(schema);
  if (schema.virtuals) {
    Object.keys(schema.virtuals).forEach(function(pathname) {
      path.push(pathname);
      handler(path.join('.'), schema.virtuals[pathname]);
      path.pop();
    });
  }
}
